8332514: Allow class space size to be larger than 3GB

tstuefe opened the pull request:
Class space size is capped at 3GB. Technically, the Klass encoding range is 4GB.

We cap class space at 3GB to leave breathing room for CDS. However, CDS needs a lot less, so we waste address space. In addition, class space is unnecessarily capped at 3GB even if we run without CDS, which is unnecessary.

What we should do instead:
- if CDS is off, use the full 4GB
- if CDS is on, use as much memory as we can after accommodating the archive

This patch implements that new behavior. 

Now, we can have 4GB class space without CDS; with CDS, a bit less (depends on the CDS archive size; by default, on MacOS, we reach about 3.98GB).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332514](https://bugs.openjdk.org/browse/JDK-8332514): Allow class space size to be larger than 3GB (**Enhancement** - P4)


### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19347/head:pull/19347` \
`$ git checkout pull/19347`

Update a local copy of the PR: \
`$ git checkout pull/19347` \
`$ git pull https://git.openjdk.org/jdk.git pull/19347/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19347`

View PR using the GUI difftool: \
`$ git pr show -t 19347`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19347.diff">https://git.openjdk.org/jdk/pull/19347.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19347#issuecomment-2127454590)

tstuefe commented:
Pinging @iklam and @coleenp

iklam commented on test/hotspot/jtreg/runtime/CompressedOops/CompressedClassSpaceSize.java:
```diff
@@ -56,20 +56,42 @@ public static void main(String[] args) throws Exception {
+
+        // With CDS, class space should fill whatever the CDS archive leaves us (modulo alignment)
+        pb = ProcessTools.createLimitedTestJavaProcessBuilder("-XX:CompressedClassSpaceSize=" + max_class_space_size,
+                "-Xshare:on", "-Xlog:metaspace*", "-version");
```
The `vm.cds` property indicates whether the JDK was built with INCLUDE_CDS. It doesn't guarantee that a CDS archive is included in the JDK -- a JDK can be configured with `--disable-cds-archive`.

It's better to generate a dummy CDS archive and use it explicitly in the test.

iklam commented on test/hotspot/jtreg/runtime/CompressedOops/CompressedClassSpaceSize.java:
```diff
@@ -25,7 +25,7 @@
  * @bug 8022865
  * @summary Tests for the -XX:CompressedClassSpaceSize command line option
- * @requires vm.bits == 64 & vm.opt.final.UseCompressedOops == true
+ * @requires vm.bits == 64 & vm.opt.final.UseCompressedOops == true & vm.cds
```
All test cases unrelated to CDS will also be disabled if the JVM is not configured with CDS.

I think it's better to move the new test cases to a new file.

tstuefe commented on src/hotspot/share/cds/metaspaceShared.cpp:
```diff
@@ -1266,6 +1266,7 @@ char* MetaspaceShared::reserve_address_space_for_archives(FileMapInfo* static_ma
   assert(static_mapinfo->mapping_base_offset() == 0, "Must be");
   size_t archive_end_offset  = (dynamic_mapinfo == nullptr) ? static_mapinfo->mapping_end_offset() : dynamic_mapinfo->mapping_end_offset();
   size_t archive_space_size = align_up(archive_end_offset, archive_space_alignment);
+  guarantee(archive_space_size < G, "weirdly large archive (" SIZE_FORMAT ")", archive_space_size);
```
I wanted to have some form of safety against e.g. malicioys attacks that modify archive size. This seemed to be the simplest way. I can exchange 1G for any other reasonable number, but think there should be a limit.

tstuefe commented:
@iklam Thanks for the review.
- I replaced the offending assert with one requiring archive size to be smaller than 4G - 16MB, for a theoretical minimal class space size of 16MB. 
- I rewrote the test. I split it into several independently executed parts, and only the part doing the dumping requires CDS. I also explicitly dump now, with an explicitly given archive.

iklam reviewed (approved):
LGTM. Thanks for making the changes.

tstuefe commented:
Thanks @iklam !

@rkennke could you take a look please? Its simple.

tstuefe commented:
Anyone else? I would like to get this into 23, and the patch is not complex. Thanks!

tstuefe commented:
Many thanks, @dholmes-ora and @iklam !

/integrate
